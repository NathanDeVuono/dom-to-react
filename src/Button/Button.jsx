import React from "react";

export default function Button({ children, className, icon, disabled }) {
  return (
    <button className={`from-component ${className}`} disabled={disabled}>
      {icon}
      {children}
    </button>
  );
}
