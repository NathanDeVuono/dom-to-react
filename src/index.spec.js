import Button from "./Button/Button";

import injectComponent, {
  attributesToProps,
  slotsToProps,
  parseChildren,
} from "./index";

describe("attributesToProps", () => {
  it("should assign all attributes to the returned object", () => {
    const node = document.createElement("button");
    node.disabled = true;
    node.classList.add("test-class");
    node.id = "test-id";

    expect(attributesToProps(node)).toMatchSnapshot();
  });
});

describe("slotsToProps", () => {
  it("should assign all slots to the returned object", () => {
    const slotOne = document.createElement("span");
    slotOne.slot = "first";
    slotOne.textContent = "First";

    const slotTwo = document.createElement("span");
    slotTwo.slot = "second";
    slotTwo.textContent = "Second";

    const node = document.createElement("div");
    node.append(slotOne);
    node.append(slotTwo);

    expect(slotsToProps(node)).toMatchSnapshot();
  });
});

describe("parseChildren", () => {
  it("should return all children as React nodes", () => {
    const childOne = document.createElement("button");
    childOne.disabled = true;
    childOne.textContent = "Don't click me";

    const subChildOne = document.createElement("em");
    subChildOne.textContent = "Sub-child";

    const childTwo = document.createElement("span");
    childTwo.id = "the-span";
    childTwo.classList.add("test");
    childTwo.textContent = "this is child two";
    childTwo.append(subChildOne);

    const node = document.createElement("div");
    node.textContent = "test string";
    node.append(childOne);
    node.append(childTwo);

    expect(parseChildren(node)).toMatchSnapshot();
  });
});

describe("injectComponent", () => {
  it("should inject a React component into DOM", () => {
    const node = document.createElement("custom-button");
    node.textContent = "Click me";

    injectComponent(Button, node);

    expect(node).toMatchSnapshot();
  });

  it("should apply props object", () => {
    const node = document.createElement("custom-button");
    node.textContent = "Click me";

    injectComponent(Button, node, {
      test: "prop",
      complex: { data: { here: true } },
      className: "extra classes",
    });

    expect(node).toMatchSnapshot();
  });
});
