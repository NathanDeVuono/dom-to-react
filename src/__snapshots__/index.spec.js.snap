// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`attributesToProps should assign all attributes to the returned object 1`] = `
Object {
  "className": "test-class",
  "disabled": true,
  "id": "test-id",
}
`;

exports[`injectComponent should apply props object 1`] = `
<custom-button>
  <button
    class="from-component extra classes"
  >
    <span>
      Click me
    </span>
  </button>
</custom-button>
`;

exports[`injectComponent should inject a React component into DOM 1`] = `
<custom-button>
  <button
    class="from-component undefined"
  >
    <span>
      Click me
    </span>
  </button>
</custom-button>
`;

exports[`parseChildren should return all children as React nodes 1`] = `
Array [
  <span>
    test string
  </span>,
  <button
    disabled={true}
  >
    <span>
      Don't click me
    </span>
  </button>,
  <span
    className="test"
    id="the-span"
  >
    <span>
      this is child two
    </span>
    <em>
      <span>
        Sub-child
      </span>
    </em>
  </span>,
]
`;

exports[`slotsToProps should assign all slots to the returned object 1`] = `
Object {
  "first": <span
    slot="first"
  >
    First
  </span>,
  "second": <span
    slot="second"
  >
    Second
  </span>,
}
`;
